### DISCUSSION TITLE
How to write Write[Tuple]

### DISCUSSION THREAD

**Discussion by nineclue** [tpolecat/doobie#1898] [Q&A] [answered]
Hi, is it possible to use scala 3's Tuple in doobie? 

What I want to do is like this
```scala
def insertStudy(study: Tuple) = 
        val sql = "INSERT INTO study VALUES (" ++ Range(0, study.size).map(_ => "?").mkString(", ") ++ ")"
        HC.prepareStatement(sql)(HPS.set(study))
```

I've got error message No given instance of type doobie.util.Write[Tuple] was found for parameter A of method set in object preparedstatement.

I think Tuple related code is in scala-3/util/WritePlatform.scala but cannot figure out how to use or write Write[Tuple].

**Comment by jatcwang** [comment, marked as answer]
Maybe this example scala-cli script will help. Note that you need to know the actual Tuple type e.g. `(Int, String)` because we need to know how to set each column at compile time.

```
//> using scala "3.3.0"
//> using dep "org.tpolecat::doobie-core:1.0.0-RC4"

import doobie.*
import doobie.implicits.*
import doobie.implicits.given

def tupleQuery[T <: Tuple](values: T)(using write: Write[T], pos: doobie.util.pos.Pos) = {
  fr"insert into study values (${write.toFragment(values)})"
}

@main
def main(): Unit = {
  Read[(Int, String)]
  Write[(Int, String)]
  tupleQuery[(Int, String, Double)]((1,"s",2.0))
}
```

**Reply by nineclue** [reply]
Thanks a lot! 
What I really trying to do is dynamic fields that is read from file at runtime. I guess it's impossible.

**Reply by jatcwang** [reply]
Yeah without knowing the types at compile time I don't think you'll gain much over using raw JDBC.
